src/mcts/search.cc

```diff
@@ -748,7 +745,10 @@ SearchWorker::NodeToProcess SearchWorker::PickNodeToExtend() {
     //            in the beginning (and there would be no need for "if
     //            (!is_root_node)"), but that would mean extra mutex lock.
     //            Will revisit that after rethinking locking strategy.
-    if (!is_root_node) node = best_edge.GetOrSpawnNode(/* parent */ node);
+    {
+      SharedMutex::Lock lock(search_->nodes_mutex_);
```

Comment:
Actually, it's fine to use atomic_flag, and in case of it being locked, return collision.
(note that it's expected that a leaf node during collision does not have n_in_flight_ incremented, they are only incremented along the path. I don't remember why it is so, probably it's better to change code that n_in_flight_ is also incremented for leaf).

In the case described above there will be some false collisions, but they will be rare.
If you want to rewrite that without possibility of false collision, it should be something like that:
```
  if not edge.NodeIsSpawned():
     if lock.test_and_set:
        return {collision}  // If we wait and follow, it will be a collision anyway, return early!
     else:
        node = edge.GetOrSpawnNode()  // May Spawn or Get.
        lock.clear()
  else:
    node = edge.GetOrSpawnNode(); // Always gets, never spawns. No lock is needed.
```

Reply:
Tested this quick patch:
```
if (!best_edge.NodeIsSpawned())  // best_edge->node_ != nullptr
    {
      SharedMutex::Lock lock(search_->nodes_mutex_);
      if (!is_root_node) node = best_edge.GetOrSpawnNode(/* parent */ node);
    } else {
      if (!is_root_node) node = best_edge.GetOrSpawnNode(/* parent */ node);
    }
```
using random backend an 8 threads, withouth patch:
`info depth 7 seldepth 14 time 15127 nodes 2797149 score cp 0 hashfull 1000 nps 184911`
with this change:
`info depth 7 seldepth 14 time 10677 nodes 2798975 score cp 0 hashfull 1000 nps 262149`

But I still don't know, do performance improvements on the random backend translate to better scaling with powerful multi-gpu setups? Or there are other issues with the efficiency of batches?

Reply:
If that is still in any way a bottleneck, it seems like an ideal case for TSX.

Reply:
Hadn't heard before of TSX but looks very interesting. I'm about to commit some changes like the ones @mooskagh proposed, testing that and it seems that now this is not the bottleneck anymore, but the load is more or less spread among other areas.

Reply:
This pseudo-code:
```
  if not edge.NodeIsSpawned():
     if lock.test_and_set:
        return {collision}  // If we wait and follow, it will be a collision anyway, return early!
     else:
        node = edge.GetOrSpawnNode()  // May Spawn or Get.
        lock.clear()
  else:
    node = edge.GetOrSpawnNode(); // Always gets, never spawns. No lock is needed.
```
Is now implemented in this branch.